File: Algorithms/Graphs/Bipartite.js
"use strict";

let bfs = (graph, vertex, colorMap) => {

    colorMap[vertex] = 1;

    let q = [vertex];

    while (q.length > 0) {

        let current = q.shift();

        for (let neighbor in graph) {

            if (graph[current][neighbor]) {

                if (colorMap[neighbor] == -1) {

                    colorMap[neighbor] = 1 - colorMap[current];

                    q.push(neighbor);

                } else if (colorMap[current] == colorMap[neighbor]) {

                    return false;

                }

            }

        }

    }

    return true;

}

let isBipartite = (graph) => {

    let colorMap = [];
    while (colorMap.push(-1) < graph.length);

    for (let i in graph) {

        if (colorMap[i] == -1 && !bfs(graph, i, colorMap)) {

            return false;

        }

    }

    return true;

}

let graph = [
    [0, 1, 1, 1],
    [1, 0, 1, 0],
    [1, 1, 0, 0],
    [1, 0, 0, 0]
];

console.log(isBipartite(graph))

File: Algorithms/Graphs/Bipartite.py

def _bfs(graph, vertex, color_map):

    color_map[vertex] = 1

    q = [vertex]

    while len(q) > 0:

        current = q.pop(0)

        for neighbor in range(len(graph)):

            if graph[current][neighbor]:

                if color_map[neighbor] == -1:

                    color_map[neighbor] = 1 - color_map[current]

                    q.append(neighbor)

                elif color_map[current] == color_map[neighbor]:

                    return False

    return True

def is_bipartite(graph):

    color_map = [ -1 ] * len(graph)

    for i in range(len(graph)):

        if color_map[i] == -1 and not _bfs(graph, i, color_map):

            return False

    return True

def main():

    graph = [
        [0, 1, 1, 1],
        [1, 0, 1, 0],
        [1, 1, 0, 0],
        [1, 0, 0, 0]
    ]

    print(is_bipartite(graph))

if __name__ == '__main__':
    main()
